src/common/function/src/scalars/timestamp/greatest.rs

```diff
@@ -0,0 +1,244 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::{self};
+use std::str::FromStr;
+
+use common_query::error::{
+    self, ArrowComputeSnafu, DowncastVectorSnafu, InvalidFuncArgsSnafu, Result,
+    UnsupportedInputDataTypeSnafu,
+};
+use common_query::prelude::{Signature, Volatility};
+use common_time::Date;
+use datatypes::arrow::array::Date32Array;
+use datatypes::arrow::compute::kernels::comparison::gt_dyn;
+use datatypes::arrow::compute::kernels::zip;
+use datatypes::prelude::ConcreteDataType;
+use datatypes::vectors::{ConstantVector, DateVector, Helper, StringVector, Vector, VectorRef};
+use snafu::{ensure, OptionExt, ResultExt};
+
+use crate::scalars::function::{Function, FunctionContext};
+
+#[derive(Clone, Debug, Default)]
+pub struct GreatestFunction;
+
+const NAME: &str = "greatest";
+
+fn convert_to_date(arg: &str) -> Option<i32> {
+    match Date::from_str(arg) {
+        Ok(ts) => Some(ts.val()),
+        Err(_err) => None,
+    }
+}
+
+fn string_vector_to_date32_array(column: &VectorRef) -> Result<Date32Array> {
+    let column = if column.is_const() {
+        let column: &ConstantVector = unsafe { Helper::static_cast(column) };
+        column
+            .inner()
+            .as_any()
+            .downcast_ref::<StringVector>()
+            .context(DowncastVectorSnafu {
+                err_msg: format!(
+                    "expect StringVector, got vector type {}",
+                    column.vector_type_name()
+                ),
+            })?
+    } else {
+        column
+            .as_any()
+            .downcast_ref::<StringVector>()
+            .context(DowncastVectorSnafu {
+                err_msg: format!(
+                    "expect StringVector, got vector type {}",
+                    column.vector_type_name()
+                ),
+            })?
+    };
+    let column = (0..column.len())
+        .map(|idx| convert_to_date(&column.get(idx).to_string()))
+        .collect::<Vec<_>>();
+    Ok(Date32Array::from_iter(column))
+}
```

Comment:
You can use https://docs.rs/arrow/latest/arrow/compute/fn.cast.html to do the cast, same to `date_vector_to_date32_array`.

I've checked it is supported https://github.com/apache/arrow-rs/blob/62bb64cf9f034a75e4485719653253077eb8efa6/arrow-cast/src/cast.rs#L187-L204

Reply:
I guess that I addressed your review: 
4e2539a